src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -561,10 +563,40 @@ void SubqueryToJoinRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node)
     return;
   }
 
-  const auto join_node = JoinNode::make(predicate_node_info->join_mode, join_predicates);
+  const auto join_mode = predicate_node_info->join_mode;
+  const auto join_node = JoinNode::make(join_mode, join_predicates);
   lqp_replace_node(node, join_node);
   join_node->set_right_input(pull_up_result.adapted_lqp);
 
+  if (pull_up_result.adapted_lqp) {
+    const auto& estimator = cost_estimator->cardinality_estimator;
+    const auto left_side_cardinality = estimator->estimate_cardinality(join_node->left_input());
+    const auto right_side_cardinality = estimator->estimate_cardinality(join_node->right_input());
+    const auto& first_join_predicate = join_predicates[0];
+
+    bool first_join_predicate_is_equals = false;
+    if (const auto predicate_expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(first_join_predicate)) {
+      if (predicate_expression->predicate_condition == PredicateCondition::Equals) {
+        first_join_predicate_is_equals = true;
+      }
+    }
+
+    if ((join_mode == JoinMode::Semi || join_mode == JoinMode::AntiNullAsTrue ||
+         join_mode == JoinMode::AntiNullAsFalse) &&
+        right_side_cardinality > left_side_cardinality * 10 && first_join_predicate_is_equals) {
+      // Semi/Anti joins are currently handled by the hash join, which performs badly if the right side is much
```

Comment:
For OLTP-type joins with very few rows, it's probably a good idea to avoid the added join? Maybe add a condition like `max(right_side_cardinality, left_side_cardinality) > 1000`?

Reply:
We want the left_side_cardinality (renamed to probe_side_cardinality) to be big. I believe the additional criterion should simply be build_side_cardinality > 1000.

Reply:
Ack.